File: ProjectCS/Data/Faculty.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace SessionResultsCS
{
    public class Faculty
    {
        public const int SPECIALITY_NOT_FOUND = -1;


        private string _name;
        // Список специальностей
        private List<Speciality> _specialities;


        // Название факультета
        public string Name
        {
            get { return _name; }
            set
            {
                string newName = value.Trim();
                if (newName.Length == 0)
                {
                    throw new ArgumentException("Faculty name cannot be empty");
                }
                _name = newName;
            }
        }


        public Faculty(string name)
        {
            Name = name;
            _specialities = new List<Speciality>();
        }


        // Возвращает количество специальностей
        public int GetSpecialityCount()
        {
            return _specialities.Count;
        }

        // Возвращает специальность по указанному индексу
        public Speciality GetSpecialityAt(int index)
        {
            return _specialities[index];
        }

        // Возвращает список специальностей
        public ReadOnlyCollection<Speciality> GetSpecialities()
        {
            return _specialities.AsReadOnly();
        }

        // Находит специальность по коду и возвращает её индекс. Если такой специальности нет в списке возвращает SPECIALITY_NOT_FOUND
        public int FindSpecialityByCode(string code)
        {
            return _specialities.FindIndex(speciality => speciality.Code == code);
        }

        // Добавляет специальность
        public void AddSpeciality(Speciality speciality)
        {
            if (FindSpecialityByCode(speciality.Code) != SPECIALITY_NOT_FOUND)
            {
                throw new ArgumentException("Faculty already contains added speciality");
            }
            _specialities.Add(speciality);
        }

        // Удаляет специальность по коду
        public void DeleteSpecialityByCode(string code)
        {
            int index = FindSpecialityByCode(code);
            if (index == SPECIALITY_NOT_FOUND)
            {
                throw new ArgumentException("Faculty doesn't contain deleted speciality");
            }
            _specialities.RemoveAt(index);
        }

        // Удаляет специальность
        public void DeleteSpeciality(Speciality speciality)
        {
            DeleteSpecialityByCode(speciality.Code);
        }
    }
}

File: Project/Data/Faculty.cpp
#include "Data/Faculty.hpp"

#include "StringUtils.hpp"

#include <stdexcept>


Faculty::Faculty(const std::string& name) {
    SetName(name);
}


std::string Faculty::ToString() const {
    return m_name;
}

const std::string& Faculty::GetName() const {
    return m_name;
}

void Faculty::SetName(const std::string& value) {
    std::string newValue = StringUtils::Trim(value);
    if (newValue.size() == 0) {
        throw std::invalid_argument("Faculty name cannot be empty");
    }
    m_name = std::move(newValue);
}

int Faculty::GetSpecialityCount() const {
    return m_specialities.size();
}

std::shared_ptr<Speciality> Faculty::GetSpecialityAt(int index) const {
    return m_specialities[index];
}

const std::vector<std::shared_ptr<Speciality>>& Faculty::GetSpecialities() const {
    return m_specialities;
}

int Faculty::FindSpecialityByCode(const std::string& code) const {
    for (int i = 0; i < m_specialities.size(); i++) {
        if (m_specialities[i]->GetCode() == code) {
            return i;
        }
    }
    return SPECIALITY_NOT_FOUND;
}


void Faculty::AddSpeciality(std::shared_ptr<Speciality> speciality) {
    if (FindSpecialityByCode(speciality->GetCode()) != SPECIALITY_NOT_FOUND) {
        throw std::invalid_argument("Faculty already contains added speciality");
    }
    if (!speciality->m_faculty.expired()) {
        throw std::runtime_error("Speciality is already added to another faculty");
    }
    m_specialities.push_back(speciality);
    speciality->m_faculty = weak_from_this();
}

void Faculty::DeleteSpecialityByCode(const std::string& code) {
    int index = FindSpecialityByCode(code);
    if (index == SPECIALITY_NOT_FOUND) {
        throw std::invalid_argument("Faculty doesn't contain deleted speciality");
    }
    m_specialities[index]->m_faculty.reset();
    m_specialities.erase(m_specialities.begin() + index);
}

void Faculty::DeleteSpeciality(std::shared_ptr<Speciality> Speciality) {
    DeleteSpecialityByCode(Speciality->GetCode());
}
